### PR TITLE
Fix generate_system_pyproject newline issues

### LINKED ISSUE #6457: Poetry's internal pyproject.toml continually grows larger with empty lines
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: Windows 11
- **Poetry version**: 1.2.0
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: N/A

## Issue

While I've been testing installing/uninstalling a plugin I wrote and running `poetry self add ...`, I've noticed that my `%APPDATA%/pypoetry/pyproject.toml` file is growing with more and more blank lines. Right now, there are 319,760 lines total. It looks like this:

```toml
[tool.poetry]
name = "poetry-instance"
version = "1.2.0"
description = ""
authors = []
license = ""

[tool.poetry.dependencies]
python = "3.10.5"
poetry = "1.2.0"

[tool.poetry.group.additional.dependencies]

# around 800 blank lines here

poetry-dynamic-versioning = {extras = ["plugin"], version = "^0.18.0"}





poetry-dynamic-versioning-plugin = "^0.4.0"

# then the rest  of the file is blank
```

### Testing from a blank slate
Delete Poetry's internal pyproject.toml and poetry.lock files.

Run: `poetry self add poetry-dynamic-versioning[plugin]`. Now pyproject.toml looks like:

```toml
[tool.poetry]
name = "poetry-instance"
version = "1.2.0"
description = ""
authors = []
license = ""

[tool.poetry.dependencies]
python = "3.10.5"
poetry = "1.2.0"

[tool.poetry.group.additional.dependencies]
poetry-dynamic-versioning = {extras = ["plugin"], version = "^0.18.0"}


```

Run: `poetry self remove poetry-dynamic-versioning`. Now pyproject.toml looks like:

```toml
[tool.poetry]
name = "poetry-instance"
version = "1.2.0"
description = ""
authors = []
license = ""

[tool.poetry.dependencies]
python = "3.10.5"
poetry = "1.2.0"

[tool.poetry.group.additional.dependencies]




```

Run: `poetry self add poetry-dynamic-versioning[plugin]`. Now pyproject.toml looks like:

```toml
[tool.poetry]
name = "poetry-instance"
version = "1.2.0"
description = ""
authors = []
license = ""

[tool.poetry.dependencies]
python = "3.10.5"
poetry = "1.2.0"

[tool.poetry.group.additional.dependencies]



poetry-dynamic-versioning = {extras = ["plugin"], version = "^0.18.0"}





```

### PR DESCRIPTION
**Pull request by NimajnebEC**
# Addressed Issues

### Additional Carriage Returns
tomlkit persists the original line separators for the preserved tables, as a result these lines will be separated by `\r\n` on WIndows. Since the `newline` kwarg is omitted from the `self.system_pyproject.write_text`<sup>[[1]](https://github.com/python-poetry/poetry/blob/71b3d6e030c879f0b4b8e3a08eac2b38a402b6bd/src/poetry/console/commands/self/self_command.py#L80)</sup> method call, all occurrences of `\n` are replaced by `os.linesep`<sup>[[2]](https://docs.python.org/3/library/functions.html#open)</sup> before writing. This causes the resultant line separators for the preserved lines to be `\r\r\n`  on Windows.

These additional carriage returns become an issue when the file is parsed by tomli, which replaces only `\r\n` with `\n`,<sup>[[3]](https://github.com/hukkin/tomli/blob/1bd3345f97cba795d7e6075956815c0a52151ed0/src/tomli/_parser.py#L74)</sup> leaving the extra return. When encountering this, the following exception will be raised: `TOMLDecodeError: Expected newline or end of document after a statement (at line 12, column 44)`.

This issue causes the following commands to prematurely exit on Windows when any plugins are installed: `self lock`, `self show`, `self update` and `self show plugins`.

### Increasing Lines at End of File
tomlkit also persists the number of blank lines after an entry, as a result when preserved tables are present, the newline added by `Factory.create_pyproject_from_package`<sup>[[4]](https://github.com/python-poetry/poetry/blob/71b3d6e030c879f0b4b8e3a08eac2b38a402b6bd/src/poetry/factory.py#L293)</sup> is unnecessary and will cause a new line to be added to the file every time the function is called.

# Changes To Fix The Issues

1. Use [`TOMLFile.write`](https://github.com/sdispater/tomlkit/blob/18a527758bfa3f8fe65e2f45f61df42f2f5af0c1/tomlkit/toml_file.py#L47) for saving the file to let tomlkit handle the translation of newlines.
2. Remove line that adds the trailing newline in [`Factory.create_pyproject_from_package`](https://github.com/python-poetry/poetry/blob/71b3d6e030c879f0b4b8e3a08eac2b38a402b6bd/src/poetry/factory.py#L293).

# Pull Request Check List
Resolves #6457 
- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

### CODE REVIEW

**Comment by dimbleby** [comment]
Something like
```patch
-        self.system_pyproject.write_text(content.as_string(), encoding="utf-8")
+        pyproject = PyProjectTOML(self.system_pyproject)
+        pyproject.file.write(content)
```
should be cleaner, let tomlkit worry about the line endings.  (Which it does, see https://github.com/sdispater/tomlkit/blob/18a527758bfa3f8fe65e2f45f61df42f2f5af0c1/tomlkit/toml_file.py#L47)

per the pipeline, `write_text()` doesn't get the `newline` parameter until python 3.10.

**Comment by NimajnebEC** [comment]
Thanks for the help! That is definitely a cleaner solution. It's just a case of finding a nice solution to prevent those newlines from being added with this new method.

**Comment by dimbleby** [comment]
well, one thing at a time.  Although the extra lines and the faulty carriage returns are in the same general area they're really quite different; and one oughtn't cause us to make a worse fix to the other.

re the unwanted newlines: do we _ever_ want that newline?  perhaps just delete the line of code that adds it altogether.  Or, make it caller responsibility to add it if they want it.

**Comment by NimajnebEC** [comment]
The newline appears to be added solely for code styling, so I'd say it is probably alright to remove. Aside from here, the `Factory.create_pyproject_from_package` method is only used in [`test_factory.py`](https://github.com/python-poetry/poetry/blob/fa5543a639e76d14e7262572ea6b06fe099c0f46/tests/test_factory.py#L152) and [`test_remove_plugins.py`](https://github.com/python-poetry/poetry/blob/fa5543a639e76d14e7262572ea6b06fe099c0f46/tests/console/commands/self/test_remove_plugins.py#L38), neither of which look like they will be effected by the removal of this newline.

File: tests/console/commands/self/test_system_pyproject.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    generated = cmd.system_pyproject.read_text()
+
+    for _i, c in enumerate(generated[::-1]):
+        if c != "\n":
+            break
+
+    assert _i == existing_newlines
+
+
+def test_generate_system_pyproject_carraige_returns(
```
Comment (radoering) [inline tests/console/commands/self/test_system_pyproject.py:43]:
```suggestion
def test_generate_system_pyproject_carriage_returns(
```

**Review by radoering — CHANGES_REQUESTED** [review]
I'd propose to rename the test file to `test_self_command.py`. Usually, we use the same name as the module we are testing.

**Comment by dimbleby** [comment]
fixes #6457, presumably

File: src/poetry/factory.py [hunk 2]
```diff
@@ -290,7 +290,6 @@ def create_pyproject_from_package(
             content["extras"] = extras_section
 
         pyproject = cast("TOMLDocument", pyproject)
-        pyproject.add(tomlkit.nl())
 
         if path:
             path.joinpath("pyproject.toml").write_text(
```
[the remark this answers is not in the document]
Comment (NimajnebEC) [reply src/poetry/factory.py:295]: Indeed, if any of the used properties contain Windows line endings then the issue will occur here too. So yeah, it is probably best to remove that parameter.

**Pull request merged**